File: web/js/pose_presets.js
/**
 * Library of pose presets — sets of joint rotations applied to an avatar's
 * skeleton. Each preset is a function `apply(joints, mirror)` that mutates
 * the joints' Euler rotations.
 *
 * mirror=true applies a left-right mirrored variant (useful when you have
 * two people facing each other).
 *
 * The classifier `pickPosePreset(personDescription)` reads the AI-produced
 * Chinese stance/upper_body/hands/gaze fields and returns a preset name.
 */

// ---------------------------------------------------------------------------
// Helpers — Three.js Euler angles in radians.
// ---------------------------------------------------------------------------

const D2R = Math.PI / 180;

function setRot(joint, x = 0, y = 0, z = 0) {
  if (!joint) return;
  joint.rotation.set(x * D2R, y * D2R, z * D2R);
}

function maybeMirror(side, mirror) {
  if (!mirror) return side;
  return side === "left" ? "right" : "left";
}

// ---------------------------------------------------------------------------
// Each preset writes into the joints map. Side-effects only.
// All angles in degrees (converted in setRot).
//
// Coordinate convention (Three.js with Y up):
//   - Shoulder rotation.z negative → arm raises sideways (away from torso)
//   - Shoulder rotation.x positive → arm reaches forward
//   - Elbow rotation.x positive → forearm bends up
//   - Hip rotation.x positive → leg lifts forward
//   - Knee rotation.x positive → calf bends back (kneeling)
// ---------------------------------------------------------------------------

const PRESETS = {
  /** Default standing, hands at side, feet shoulder-width. */
  standing(joints) {
    setRot(joints.leftShoulder, 0, 0, -78);
    setRot(joints.rightShoulder, 0, 0, 78);
    setRot(joints.leftElbow, 5, 0, 0);
    setRot(joints.rightElbow, 5, 0, 0);
    setRot(joints.leftHip, 0, 0, 0);
    setRot(joints.rightHip, 0, 0, 0);
    setRot(joints.leftKnee, 0, 0, 0);
    setRot(joints.rightKnee, 0, 0, 0);
    setRot(joints.head, -2, 0, 0);
    setRot(joints.neck, 0, 0, 0);
  },

  /** Hands clasped in front, gentle smile. */
  hands_clasped(joints) {
    setRot(joints.leftShoulder, 0, 10, -50);
    setRot(joints.rightShoulder, 0, -10, 50);
    setRot(joints.leftElbow, 90, 0, 0);
    setRot(joints.rightElbow, 90, 0, 0);
    setRot(joints.leftForearm, 0, -25, 0);
    setRot(joints.rightForearm, 0, 25, 0);
    setRot(joints.head, 0, 0, 0);
  },

  /** Walking stride, opposite arm/leg swing. */
  walking(joints, mirror) {
    const fwd = mirror ? -1 : 1;
    setRot(joints.leftShoulder, -25 * fwd, 0, -75);
    setRot(joints.rightShoulder, 25 * fwd, 0, 75);
    setRot(joints.leftElbow, 25, 0, 0);
    setRot(joints.rightElbow, 25, 0, 0);
    setRot(joints.leftHip, 22 * fwd, 0, 0);
    setRot(joints.rightHip, -22 * fwd, 0, 0);
    setRot(joints.leftKnee, 12, 0, 0);
    setRot(joints.rightKnee, 0, 0, 0);
    setRot(joints.head, 5, mirror ? 8 : -8, 0);
  },

  /** Half squat sitting on a low surface. */
  half_sit(joints) {
    setRot(joints.leftHip, 70, 6, 0);
    setRot(joints.rightHip, 70, -6, 0);
    setRot(joints.leftKnee, -90, 0, 0);
    setRot(joints.rightKnee, -90, 0, 0);
    setRot(joints.leftShoulder, 0, 5, -60);
    setRot(joints.rightShoulder, 0, -5, 60);
    setRot(joints.leftElbow, 65, 0, 0);
    setRot(joints.rightElbow, 65, 0, 0);
  },

  /** Crouching low. */
  crouch(joints) {
    setRot(joints.leftHip, 95, 8, 0);
    setRot(joints.rightHip, 95, -8, 0);
    setRot(joints.leftKnee, -130, 0, 0);
    setRot(joints.rightKnee, -130, 0, 0);
    setRot(joints.leftShoulder, 25, 0, -50);
    setRot(joints.rightShoulder, 25, 0, 50);
    setRot(joints.leftElbow, 50, 0, 0);
    setRot(joints.rightElbow, 50, 0, 0);
    setRot(joints.head, 6, 0, 0);
  },

  /** Looking back over shoulder. */
  looking_back(joints, mirror) {
    setRot(joints.head, -3, mirror ? -110 : 110, 0);
    setRot(joints.neck, 0, mirror ? -25 : 25, 0);
    setRot(joints.torso, 0, mirror ? -8 : 8, 0);
    // Hands lightly held in front
    setRot(joints.leftShoulder, 0, 10, -55);
    setRot(joints.rightShoulder, 0, -10, 55);
    setRot(joints.leftElbow, 70, 0, 0);
    setRot(joints.rightElbow, 70, 0, 0);
  },

  /** Holding hands sideways (use side='left' = inner hand). */
  holding_hands(joints, mirror) {
    // Inner arm extended out, outer arm relaxed
    const inner = mirror ? "right" : "left";
    const outer = mirror ? "left" : "right";
    setRot(joints[`${inner}Shoulder`], 5, 0, mirror ? 70 : -70);
    setRot(joints[`${inner}Elbow`], 12, 0, 0);
    setRot(joints[`${outer}Shoulder`], 0, 0, mirror ? -75 : 75);
    setRot(joints[`${outer}Elbow`], 5, 0, 0);
    setRot(joints.head, -3, mirror ? -8 : 8, 0);
  },

  /** Hand on hip, the other relaxed, slight contrapposto. */
  hand_on_hip(joints, mirror) {
    const side = mirror ? "right" : "left";
    const other = mirror ? "left" : "right";
    setRot(joints[`${side}Shoulder`], 0, 0, mirror ? 30 : -30);
    setRot(joints[`${side}Elbow`], 100, 0, 0);
    setRot(joints[`${other}Shoulder`], 0, 0, mirror ? -78 : 78);
    setRot(joints[`${other}Elbow`], 5, 0, 0);
    setRot(joints.torso, 0, 0, mirror ? 4 : -4);
    setRot(joints.head, -2, mirror ? 6 : -6, 0);
  },

  /** V-sign with right hand (or left if mirrored), other hand at side. */
  v_sign(joints, mirror) {
    const v = mirror ? "left" : "right";
    const other = mirror ? "right" : "left";
    setRot(joints[`${v}Shoulder`], -50, 0, mirror ? -45 : 45);
    setRot(joints[`${v}Elbow`], 95, 0, 0);
    setRot(joints[`${other}Shoulder`], 0, 0, mirror ? 78 : -78);
    setRot(joints[`${other}Elbow`], 5, 0, 0);
    setRot(joints.head, -2, mirror ? -4 : 4, 0);
  },

  /** Arms crossed in front. */
  arms_crossed(joints) {
    setRot(joints.leftShoulder, 0, 30, -50);
    setRot(joints.rightShoulder, 0, -30, 50);
    setRot(joints.leftElbow, 95, 0, 0);
    setRot(joints.rightElbow, 95, 0, 0);
    setRot(joints.leftForearm, 0, 35, 0);
    setRot(joints.rightForearm, 0, -35, 0);
    setRot(joints.head, -2, 0, 0);
  },

  /** Looking at partner with subtle lean (use mirror to face the other side). */
  facing_partner(joints, mirror) {
    setRot(joints.torso, 0, mirror ? -18 : 18, 0);
    setRot(joints.head, -3, mirror ? -25 : 25, 0);
    setRot(joints.leftShoulder, 0, 5, -65);
    setRot(joints.rightShoulder, 0, -5, 65);
    setRot(joints.leftElbow, 35, 0, 0);
    setRot(joints.rightElbow, 35, 0, 0);
  },

  /** Leaning against something (slight tilt + hand back). */
  leaning(joints, mirror) {
    setRot(joints.torso, -3, 0, mirror ? 6 : -6);
    setRot(joints.leftShoulder, mirror ? -30 : 0, 0, mirror ? 30 : -78);
    setRot(joints.rightShoulder, mirror ? 0 : -30, 0, mirror ? 78 : -30);
    setRot(joints.leftElbow, mirror ? 25 : 5, 0, 0);
    setRot(joints.rightElbow, mirror ? 5 : 25, 0, 0);
    setRot(joints.head, 0, mirror ? -10 : 10, 0);
  },
};

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

export function listPosePresets() {
  return Object.keys(PRESETS);
}

export function applyPosePreset(name, joints, opts = {}) {
  const fn = PRESETS[name] || PRESETS.standing;
  // Always start from a relaxed default before stacking the preset, so a
  // pose with fewer joints set still looks natural (e.g. legs stay rest).
  PRESETS.standing(joints);
  fn(joints, opts.mirror === true);
  if (opts.bowHead) {
    setRot(joints.head, joints.head.rotation.x * (180 / Math.PI) - 6, 0, 0);
  }
}

/**
 * Heuristic mapping from a single AI-generated PersonPose's text fields
 * (stance, upper_body, hands, gaze) to a preset name. Plays it safe and
 * defaults to "standing" when nothing matches.
 */
export function pickPosePreset(person) {
  const blob = [
    person?.stance || "",
    person?.upper_body || "",
    person?.hands || "",
    person?.gaze || "",
    person?.position_hint || "",
  ].join(" ").toLowerCase();

  // Order matters — most specific patterns first.
  if (/v\s?字|v[-\s]?sign|比耶|比v|peace/i.test(blob)) return "v_sign";
  if (/牵手|拉手|hold(?:ing)?\s+hand/i.test(blob)) return "holding_hands";
  if (/回头|看向|看着|gaze|turn(?:ing)?\s+(?:back|head)|over\s+shoulder/i.test(blob)) return "looking_back";
  if (/抱臂|交叉|cross(?:ed)?\s+arms/i.test(blob)) return "arms_crossed";
  if (/叉腰|手插腰|hand\s+on\s+hip/i.test(blob)) return "hand_on_hip";
  if (/坐|sit|半坐/i.test(blob)) return "half_sit";
  if (/蹲|crouch|squat|半蹲/i.test(blob)) return "crouch";
  if (/靠|lean(?:ing)?/i.test(blob)) return "leaning";
  if (/走|漫步|散步|walk(?:ing)?|stride|前行|向前/i.test(blob)) return "walking";
  if (/对视|面向|相对|face|看向\s*person/i.test(blob)) return "facing_partner";
  if (/双手|交握|clasp/i.test(blob)) return "hands_clasped";
  return "standing";
}

/**
 * Map AI expression keyword (Chinese / English) to one of our 5 supported
 * face states.
 */
export function classifyExpression(person) {
  const t = (person?.expression || "").toLowerCase();
  if (!t) return "neutral";
  if (/抿嘴|smirk|淡笑|微微一笑/i.test(t)) return "smirk";
  if (/惊|surprised|wide-?eyed|睁大|意外/i.test(t)) return "surprised";
  if (/认真|沉思|皱眉|pensive|思|frown/i.test(t)) return "pensive";
  if (/笑|smile|joy|happy|开心|愉悦|轻松/i.test(t)) return "joy";
  return "neutral";
}

File: ios/AIPhotoCoach/Features/ARGuide/PosePresets.swift
import SceneKit

/// Swift port of web/js/pose_presets.js.
///
/// Each preset rotates a subset of the joints produced by
/// AvatarBuilderSCN. We keep the joint *names* identical to the JS
/// builder so the same AI-rationale-driven preset names work on both
/// clients.
enum PosePresets {
static let all: [String] = [
        "standing", "hands_clasped", "walking", "half_sit", "crouch",
        "looking_back", "holding_hands", "hand_on_hip", "v_sign",
    "arms_crossed", "facing_partner", "leaning", "pockets",
    "shy_pose", "hands_back", "wave",
    ]

static func apply(_ name: String, joints: [String: SCNNode], mirror: Bool = false) {
        // Always start from standing so any unset joint stays sensible.
        standing(joints)
        switch name {
        case "hands_clasped": handsClasped(joints)
        case "walking": walking(joints, mirror: mirror)
        case "half_sit": halfSit(joints)
        case "crouch": crouch(joints)
        case "looking_back": lookingBack(joints, mirror: mirror)
        case "holding_hands": holdingHands(joints, mirror: mirror)
        case "hand_on_hip": handOnHip(joints, mirror: mirror)
        case "v_sign": vSign(joints, mirror: mirror)
        case "arms_crossed": armsCrossed(joints)
        case "facing_partner": facingPartner(joints, mirror: mirror)
        case "leaning": leaning(joints, mirror: mirror)
        case "pockets": pockets(joints, mirror: mirror)
        case "shy_pose": shyPose(joints, mirror: mirror)
        case "hands_back": handsBack(joints, mirror: mirror)
        case "wave": wave(joints, mirror: mirror)
        default: break // standing already applied
        }
    }

static func pick(for person: PersonPose) -> String {
        let blob = [
            person.stance ?? "",
            person.upperBody ?? "",
            person.hands ?? "",
            person.gaze ?? "",
            person.positionHint ?? "",
        ].joined(separator: " ").lowercased()

        if blob.range(of: #"v\s?字|v[-\s]?sign|比耶|比v|peace"#,
                      options: .regularExpression) != nil { return "v_sign" }
    if blob.range(of: #"挥手|招手|wave|hello|打招呼"#,
              options: .regularExpression) != nil { return "wave" }
    if blob.range(of: #"插兜|口袋|pocket"#,
              options: .regularExpression) != nil { return "pockets" }
    if blob.range(of: #"背手|双手背后|手放身后|behind\s+back"#,
              options: .regularExpression) != nil { return "hands_back" }
    if blob.range(of: #"害羞|甜美|可爱|少女|提裙|轻提裙摆|shy|cute|sweet"#,
              options: .regularExpression) != nil { return "shy_pose" }
        if blob.range(of: #"牵手|拉手|hold(?:ing)?\s+hand"#,
                      options: .regularExpression) != nil { return "holding_hands" }
        if blob.range(of: #"回头|看向|看着|gaze|over\s+shoulder"#,
                      options: .regularExpression) != nil { return "looking_back" }
        if blob.range(of: #"抱臂|交叉|cross(?:ed)?\s+arms"#,
                      options: .regularExpression) != nil { return "arms_crossed" }
        if blob.range(of: #"叉腰|手插腰|hand\s+on\s+hip"#,
                      options: .regularExpression) != nil { return "hand_on_hip" }
        if blob.range(of: #"坐|sit|半坐"#,
                      options: .regularExpression) != nil { return "half_sit" }
        if blob.range(of: #"蹲|crouch|squat|半蹲"#,
                      options: .regularExpression) != nil { return "crouch" }
        if blob.range(of: #"靠|lean(?:ing)?"#,
                      options: .regularExpression) != nil { return "leaning" }
        if blob.range(of: #"走|漫步|散步|walk(?:ing)?|stride|前行|向前"#,
                      options: .regularExpression) != nil { return "walking" }
        if blob.range(of: #"对视|面向|相对|face"#,
                      options: .regularExpression) != nil { return "facing_partner" }
        if blob.range(of: #"双手|交握|clasp"#,
                      options: .regularExpression) != nil { return "hands_clasped" }
        return "standing"
    }

static func classifyExpression(_ person: PersonPose) -> ExpressionRenderer.Expression {
        let t = (person.expression ?? "").lowercased()
        if t.isEmpty { return .neutral }
        if t.range(of: #"抿嘴|smirk|淡笑|微微一笑"#, options: .regularExpression) != nil { return .smirk }
        if t.range(of: #"惊|surprised|wide-?eyed|睁大|意外"#, options: .regularExpression) != nil { return .surprised }
        if t.range(of: #"认真|沉思|皱眉|pensive|frown"#, options: .regularExpression) != nil { return .pensive }
        if t.range(of: #"笑|smile|joy|happy|开心|愉悦|轻松"#, options: .regularExpression) != nil { return .joy }
        return .neutral
    }

    // ---- presets --------------------------------------------------------------

    private static let d2r: Float = .pi / 180

    private static func setEuler(_ j: SCNNode?, x: Float = 0, y: Float = 0, z: Float = 0) {
        j?.eulerAngles = SCNVector3(x * d2r, y * d2r, z * d2r)
    }

    private static func standing(_ j: [String: SCNNode]) {
        setEuler(j["leftShoulder"], x: 6, y: 4, z: -10)
        setEuler(j["rightShoulder"], x: 6, y: -4, z: 10)
        setEuler(j["leftElbow"], x: 12)
        setEuler(j["rightElbow"], x: 12)
        setEuler(j["leftForearm"], y: -4)
        setEuler(j["rightForearm"], y: 4)
        setEuler(j["leftHip"]); setEuler(j["rightHip"])
        setEuler(j["leftKnee"]); setEuler(j["rightKnee"])
        setEuler(j["head"], x: -1)
        setEuler(j["neck"])
    }

    private static func pockets(_ j: [String: SCNNode], mirror: Bool) {
        let pocket = mirror ? "right" : "left"
        let free = mirror ? "left" : "right"
        setEuler(j["\(pocket)Shoulder"], x: 10, y: mirror ? -6 : 6, z: mirror ? 18 : -18)
        setEuler(j["\(pocket)Elbow"], x: 38)
        setEuler(j["\(pocket)Forearm"], y: mirror ? -14 : 14)
        setEuler(j["\(free)Shoulder"], x: 4, y: mirror ? 2 : -2, z: mirror ? -8 : 8)
        setEuler(j["\(free)Elbow"], x: 10)
        setEuler(j["torso"], z: mirror ? 4 : -4)
        setEuler(j["head"], x: -2, y: mirror ? -8 : 8)
    }

    private static func shyPose(_ j: [String: SCNNode], mirror: Bool) {
        setEuler(j["leftShoulder"], x: 8, y: 8, z: -24)
        setEuler(j["rightShoulder"], x: 8, y: -8, z: 24)
        setEuler(j["leftElbow"], x: 72)
        setEuler(j["rightElbow"], x: 72)
        setEuler(j["leftForearm"], y: -12)
        setEuler(j["rightForearm"], y: 12)
        setEuler(j["torso"], x: 2, z: mirror ? 5 : -5)
        setEuler(j["head"], x: 6, y: mirror ? -12 : 12)
    }

    private static func handsBack(_ j: [String: SCNNode], mirror: Bool) {
        setEuler(j["leftShoulder"], x: -16, y: -10, z: mirror ? 8 : -8)
        setEuler(j["rightShoulder"], x: -16, y: 10, z: mirror ? -8 : 8)
        setEuler(j["leftElbow"], x: 28)
        setEuler(j["rightElbow"], x: 28)
        setEuler(j["leftForearm"], y: -18)
        setEuler(j["rightForearm"], y: 18)
        setEuler(j["head"], x: -2, y: mirror ? -6 : 6)
    }

    private static func wave(_ j: [String: SCNNode], mirror: Bool) {
        let waving = mirror ? "left" : "right"
        let free = mirror ? "right" : "left"
        setEuler(j["\(waving)Shoulder"], x: -42, z: mirror ? -36 : 36)
        setEuler(j["\(waving)Elbow"], x: 94)
        setEuler(j["\(waving)Forearm"], y: mirror ? -10 : 10)
        setEuler(j["\(free)Shoulder"], x: 4, z: mirror ? 10 : -10)
        setEuler(j["\(free)Elbow"], x: 10)
        setEuler(j["head"], x: -1, y: mirror ? -10 : 10)
    }

    private static func handsClasped(_ j: [String: SCNNode]) {
        setEuler(j["leftShoulder"], y: 10, z: -50)
        setEuler(j["rightShoulder"], y: -10, z: 50)
        setEuler(j["leftElbow"], x: 90)
        setEuler(j["rightElbow"], x: 90)
        setEuler(j["leftForearm"], y: -25)
        setEuler(j["rightForearm"], y: 25)
    }

    private static func walking(_ j: [String: SCNNode], mirror: Bool) {
        let f: Float = mirror ? -1 : 1
        setEuler(j["leftShoulder"], x: -25 * f, z: -75)
        setEuler(j["rightShoulder"], x: 25 * f, z: 75)
        setEuler(j["leftElbow"], x: 25)
        setEuler(j["rightElbow"], x: 25)
        setEuler(j["leftHip"], x: 22 * f)
        setEuler(j["rightHip"], x: -22 * f)
        setEuler(j["leftKnee"], x: 12)
        setEuler(j["head"], x: 5, y: mirror ? 8 : -8)
    }

    private static func halfSit(_ j: [String: SCNNode]) {
        setEuler(j["leftHip"], x: 70, y: 6)
        setEuler(j["rightHip"], x: 70, y: -6)
        setEuler(j["leftKnee"], x: -90); setEuler(j["rightKnee"], x: -90)
        setEuler(j["leftShoulder"], y: 5, z: -60)
        setEuler(j["rightShoulder"], y: -5, z: 60)
        setEuler(j["leftElbow"], x: 65); setEuler(j["rightElbow"], x: 65)
    }

    private static func crouch(_ j: [String: SCNNode]) {
        setEuler(j["leftHip"], x: 95, y: 8); setEuler(j["rightHip"], x: 95, y: -8)
        setEuler(j["leftKnee"], x: -130); setEuler(j["rightKnee"], x: -130)
        setEuler(j["leftShoulder"], x: 25, z: -50); setEuler(j["rightShoulder"], x: 25, z: 50)
        setEuler(j["leftElbow"], x: 50); setEuler(j["rightElbow"], x: 50)
        setEuler(j["head"], x: 6)
    }

    private static func lookingBack(_ j: [String: SCNNode], mirror: Bool) {
        setEuler(j["head"], x: -3, y: mirror ? -110 : 110)
        setEuler(j["neck"], y: mirror ? -25 : 25)
        setEuler(j["torso"], y: mirror ? -8 : 8)
        setEuler(j["leftShoulder"], y: 10, z: -55); setEuler(j["rightShoulder"], y: -10, z: 55)
        setEuler(j["leftElbow"], x: 70); setEuler(j["rightElbow"], x: 70)
    }

    private static func holdingHands(_ j: [String: SCNNode], mirror: Bool) {
        let inner = mirror ? "right" : "left"
        let outer = mirror ? "left" : "right"
        setEuler(j["\(inner)Shoulder"], x: 5, z: mirror ? 70 : -70)
        setEuler(j["\(inner)Elbow"], x: 12)
        setEuler(j["\(outer)Shoulder"], z: mirror ? -75 : 75)
        setEuler(j["\(outer)Elbow"], x: 5)
        setEuler(j["head"], x: -3, y: mirror ? -8 : 8)
    }

    private static func handOnHip(_ j: [String: SCNNode], mirror: Bool) {
        let side = mirror ? "right" : "left"
        let other = mirror ? "left" : "right"
        setEuler(j["\(side)Shoulder"], z: mirror ? 30 : -30)
        setEuler(j["\(side)Elbow"], x: 100)
        setEuler(j["\(other)Shoulder"], z: mirror ? -78 : 78)
        setEuler(j["\(other)Elbow"], x: 5)
        setEuler(j["torso"], z: mirror ? 4 : -4)
        setEuler(j["head"], x: -2, y: mirror ? 6 : -6)
    }

    private static func vSign(_ j: [String: SCNNode], mirror: Bool) {
        let v = mirror ? "left" : "right"
        let other = mirror ? "right" : "left"
        setEuler(j["\(v)Shoulder"], x: -50, z: mirror ? -45 : 45)
        setEuler(j["\(v)Elbow"], x: 95)
        setEuler(j["\(other)Shoulder"], z: mirror ? 78 : -78)
        setEuler(j["\(other)Elbow"], x: 5)
        setEuler(j["head"], x: -2, y: mirror ? -4 : 4)
    }

    private static func armsCrossed(_ j: [String: SCNNode]) {
        setEuler(j["leftShoulder"], y: 30, z: -50)
        setEuler(j["rightShoulder"], y: -30, z: 50)
        setEuler(j["leftElbow"], x: 95); setEuler(j["rightElbow"], x: 95)
        setEuler(j["leftForearm"], y: 35); setEuler(j["rightForearm"], y: -35)
    }

    private static func facingPartner(_ j: [String: SCNNode], mirror: Bool) {
        setEuler(j["torso"], y: mirror ? -18 : 18)
        setEuler(j["head"], x: -3, y: mirror ? -25 : 25)
        setEuler(j["leftShoulder"], y: 5, z: -65); setEuler(j["rightShoulder"], y: -5, z: 65)
        setEuler(j["leftElbow"], x: 35); setEuler(j["rightElbow"], x: 35)
    }

    private static func leaning(_ j: [String: SCNNode], mirror: Bool) {
        setEuler(j["torso"], x: -3, z: mirror ? 6 : -6)
        setEuler(j["leftShoulder"], x: mirror ? -30 : 0, z: mirror ? 30 : -78)
        setEuler(j["rightShoulder"], x: mirror ? 0 : -30, z: mirror ? 78 : -30)
        setEuler(j["head"], y: mirror ? -10 : 10)
    }
}
